src/util/ConfigManager/ConfigManager.cpp

```diff
@@ -34,24 +34,70 @@
 #include "util/json.h"
 
 namespace ad_utility {
-static auto configurationOptionsImpl(auto& configurationOptions) {
-  return std::views::transform(configurationOptions, [](auto& pair) {
-    // Make sure, that there is no null pointer.
-    AD_CORRECTNESS_CHECK(pair.second);
+// ____________________________________________________________________________
+template <typename ReturnPointer>
+std::vector<std::pair<std::string, ReturnPointer>>
+ConfigManager::configurationOptionsImpl(auto& configurationOptions,
+                                        std::string_view pathPrefix) {
+  std::vector<std::pair<std::string, ReturnPointer>> collectedOptions;
 
-    // Return a dereferenced reference.
-    return std::tie(pair.first, *pair.second);
-  });
+  std::ranges::for_each(
+      configurationOptions,
+      [&collectedOptions, &pathPrefix](auto pair) {
+        const std::string& pathToCurrentEntry =
+            absl::StrCat(pathPrefix, std::get<0>(pair));
+
+        std::visit(
+            [&collectedOptions, &pathToCurrentEntry]<typename T>(T& var) {
+              // A normal `ConfigOption` can be directly added. For a
+              // `ConfigManager` we have to recursively collect the options.
+              if constexpr (std::is_same_v<std::decay_t<T>, ConfigOption>) {
+                collectedOptions.push_back(
+                    std::make_pair(pathToCurrentEntry, &var));
+              } else {
+                AD_CORRECTNESS_CHECK(
+                    (std::is_same_v<std::decay_t<T>, ConfigManager>));
+                ad_utility::appendVector(
+                    collectedOptions,
+                    configurationOptionsImpl<ReturnPointer>(
+                        var.configurationOptions_, pathToCurrentEntry));
+              }
+            },
+            std::get<1>(pair));
+      },
+      [&pathPrefix](auto& pair) {
+        // Make sure, that there is no null pointer.
+        AD_CORRECTNESS_CHECK(pair.second != nullptr);
+
+        // An empty sub manager tends to point to a logic error on the user
+        // side.
+        if (const ConfigManager* ptr =
+                std::get_if<ConfigManager>(pair.second.get());
+            ptr != nullptr && ptr->configurationOptions_.empty()) {
+          throw std::runtime_error(absl::StrCat(
+              "The sub manager at '", pathPrefix, std::get<0>(pair),
+              "' is empty. Either fill it, or delete it."));
+        }
+
```

Comment:
I find those two lambdas inside the `for_each` a little bit too long.
Maybe factor them both out (or maybe the second one is enough) and give it a proper name like
`checkNonEmptyAndDereference` and pass that as the second argument to your `for_each`. Or for symmetry also do the same for the first lambda.